### PR TITLE
Better error logging

### LINKED ISSUE #1511: Better error messages
We need better logging on the server side that logs errors that occur along with providing some context so that we can debug these errors for clients.

### PR DESCRIPTION
**Pull request by gideonthomas**
Fix #1511

This patch implements better error logging practices as well as makes use of the work done in #1506 to display better error messages to the user.

This patch is divided into 4 distinct commits outlined as follows:

1. A quick optimization fix
2. The initial foundation for enabling logging which includes:
  * Server request logging with morgan
  * A logging module that can be used to log custom error messages
1. Adds the error handlers to the server to catch 404s and other errors, send them to the client and also log them
4. Calls the error handler implemented above for all error points in every route's logic.
5. JShint error fixes

### CODE REVIEW

File: server/index.js [hunk 1]
```diff
@@ -13,13 +13,14 @@ let templatize = require("./templatize");
...
-let errorhandling = require("./lib/errorhandling");
+const HttpError = require("./lib/http-error.js");
 let middleware = require("./lib/middleware")();
 let routes = require("./routes")();
-let getFileList = require("./lib/utils").getFileList;
+const Utils = require("./lib/utils");
 
 let server = express();
-let isDevelopment = env.get("NODE_ENV") === "development";
+const environment = env.get("NODE_ENV");
```
Comment (cadecairos) [inline server/index.js]: It feels strange to use `const` all of a sudden, while values around it could technically also be `const`.

Comment (gideonthomas) [inline server/index.js]: around it?

Comment (cadecairos) [inline server/index.js]: yes, for example:

`let server = express()`

is server ever reassigned? probably not, so why isn't it const? it seems the current style in this file is using let for variable declaration.

Comment (gideonthomas) [inline server/index.js]: should I use that? Yeah it's pretty inconsistent right now. The idea is to do a full es6 re-write once node 6 is lts (october); so this is just to make that patch a bit smaller.

Comment (cadecairos) [inline server/index.js]: I'm not sure if making the code style inconsistent now is better than changing a few more lets to consts later on.

Comment (gideonthomas) [inline server/index.js]: ok, I'll make that change

**Comment by cadecairos** [comment]
"500: Failed to get the list of your projects! The request to the publish server to login the user could not be completed. "

This first and last half of this message don't seem related?

**Comment by gideonthomas** [comment]
@cadecairos updated the PR with the following:
* The message now shown to the user will be something like "Service Unavailable (500): Failed to create the project!" for e.g. which is cleaner and allows for fewer l10n strings
* To do the above, I had to change it so that the error message for the request is set at the middleware level. That is to say, an error message to use is already set for a "request to create a file" regardless of how many sub requests is needed to do that.
* I followed whatever convention is followed in the file, i.e. if `var` is used use `var`, if `let` is used use `let`, if it is a new file use `const`.

Based on this, re-r?

**Comment by gideonthomas** [comment]
@cadecairos re-ping for review

**Comment by cadecairos** [comment]
R+

**Pull request merged**